src/pyclass/create_type_object.rs

```diff
@@ -32,22 +32,37 @@ where
     T: PyClass,
 {
     unsafe {
-        PyTypeBuilder::default()
-            .type_doc(T::doc(py)?)
-            .offsets(T::dict_offset(), T::weaklist_offset())
-            .slot(ffi::Py_tp_base, T::BaseType::type_object_raw(py))
-            .slot(ffi::Py_tp_dealloc, tp_dealloc::<T> as *mut c_void)
-            .set_is_basetype(T::IS_BASETYPE)
-            .set_is_mapping(T::IS_MAPPING)
-            .set_is_sequence(T::IS_SEQUENCE)
-            .class_items(T::items_iter())
-            .build(py, T::NAME, T::MODULE, std::mem::size_of::<T::Layout>())
+        PyTypeBuilder {
+            slots: Vec::new(),
+            method_defs: Vec::new(),
+            getset_builders: HashMap::new(),
+            cleanup: Vec::new(),
+            tp_base: T::BaseType::type_object_raw(py),
+            tp_dealloc: tp_dealloc::<T>,
+            tp_dealloc_with_gc: tp_dealloc_with_gc::<T>,
```

Comment:
The implementation here compiles both `tp_dealloc` and `tp_dealloc_with_gc` for all `T`, and then selects which to actually use based on whether `has_traverse` gets set later. 

Given that there's a one-line difference between the two functions I didn't think it was worth trying to reduce the minimal code bloat by doing something fancy (probably with autoref specialization) to work out which one to use at compile time.

Reply:
I also think having both fields in `PyTypeBuilder` is not an issue. I do wonder about inlining though, i.e. whether forwarding from `tp_dealloc_with_gc` to `tp_dealloc` is good.

I suspect both functions are relatively unlikely being inlined as they are called indirectly. So maybe should call `crate::impl_::trampoline::dealloc(obj, PyCell::<T>::tp_dealloc)` directyl from within `tp_dealloc_with_gc` as well?

Reply:
Sorry for the long delay, just force-pushed to adjust `tp_alloc_with_gc` as suggested.